source/common/common/mem_block_builder.h

```diff
@@ -0,0 +1,126 @@
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+
+// Manages a block of raw memory for objects of type T. T must be
+// empty-constructible. This class carries extra member variables for tracking
+// size, and a write-pointer to support safe appends.
```

Comment:
This will default construct Ts in the count of `size`, even if you'd never use that space. That has overhead you don't actually need, right? You actually memcpy into this so whatever work is required to default construct T is entirely wasted. And if you never use that capacity you're paying for default construction you don't need.